python/ray/data/dataset.py

```diff
@@ -4614,14 +4613,16 @@ def to_pandas(self, limit: int = None) -> "pandas.DataFrame":
                     f"{count} rows will fit in local memory, set "
                     "ds.to_pandas(limit=None) to disable limits."
                 )
-        bundles = self.iter_internal_ref_bundles()
-        output = DelegatingBlockBuilder()
 
-        for bundle in bundles:
-            for block_ref in bundle.block_refs:
-                output.add_block(ray.get(block_ref))
-        block = output.build()
-        return _block_to_df(block)
+        builder = PandasBlockBuilder()
+        for batch in self.iter_batches(batch_format="pandas", batch_size=None):
+            builder.add_block(batch)
+        block = builder.build()
+
+        # `PandasBlockBuilder` creates a dataframe with internal extension types like
+        # 'TensorDtype'. We use the `to_pandas` method to convert these extension
+        # types to regular types.
+        return BlockAccessor.for_block(block).to_pandas()
```

Comment:
if i'm understanding correctly, i think this new code now adds an extra conversion from Arrow to pandas batch. is there any way we can avoid it?

the old code:
Arrow internal -> Arrow Block -> build to big Arrow block -> convert to pd DF

the new code:
**Arrow internal -> Pandas batch** -> build to big Arrow block -> convert to pd DF

Reply:
There's no step 3 ("build to big Arrow block") in the new code, since we're using PandasBlockBuilder

Reply:
ah you're right, i mistakenly thought it converts to arrow tables.